added LabelSpacing, MultiLabel and LabelStep properties to contours series

UndisputedBasement33 opened the pull request:
Fixes #1511 .

### Checklist

- [x] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- implemented LabelSpacing and LabelStep to contour series
- in order to not change default behaviour added property MultiLabel, to enable several labels per contour to make use of LabelSpacing
-

@oxyplot/admins

VisualMelon commented:
I'm still getting my head around this, but could you run a Style-Cop pass and fix the issues? It's little things like spaces at the start of comments, and expanding `if () continue;` onto multiple lines for clarity.

VisualMelon commented on Source/OxyPlot/Series/ContourSeries.cs:
```diff
@@ -123,6 +125,12 @@ public OxyColor ActualColor
         public double LabelSpacing { get; set; }
 
+        /// <summary>
+        /// Gets or sets if multiple labels should get displayed per Contour.
```
nit: "get" -> "be"

VisualMelon commented on Source/OxyPlot/Series/ContourSeries.cs:
```diff
@@ -46,8 +47,9 @@ public ContourSeries()
-            this.LabelStep = 1;
+            this.MultiLabel = false;
+            this.LabelSpacing = 150;
+            this.LabelStep = 0;
```
Having `LabelStep == 0` seems to have some dedicated behaviour: could this be explained in the inline documentation?

VisualMelon commented on Source/OxyPlot/Series/ContourSeries.cs:
```diff
@@ -11,6 +11,7 @@ namespace OxyPlot.Series
 {
     using System;
     using System.Collections.Generic;
+    using System.Diagnostics;
```
Not needed

VisualMelon commented on Source/OxyPlot/Series/ContourSeries.cs:
```diff
@@ -277,26 +285,98 @@ public override void Render(IRenderContext rc)
+                    //find index of contour points where next label should be positioned
+                    for (int k = intervalIndex; k < transformedPoints.Length; k++)
+                    {
+                        contourPartLength += Math.Sqrt(Math.Pow(transformedPoints[k].x - transformedPoints[k - 1].x, 2) + Math.Pow(transformedPoints[k].y - transformedPoints[k - 1].y, 2));
```
Could be 

                        contourPartLength += (transformedPoints[k] - transformedPoints[k - 1]).Length;

For clarity

UndisputedBasement33 commented:
Hi,
thanks for reviewing. I run through style cop and took care about your other points.

Jonarw commented:
I am also not too familiar with the `ContourSeries`, but something does not seem completely right to me.

From the the way the `LabelStep` property is described, I would expect that it it set `LabelStep=2`, every second contour will have a label. This is true as long as the `ContourLevelStep==1` (which is the case for all examples in the ExampleLibrary).

But if I set e.g. `ContourLevelStep==0.5`, I will only have a label every fourth contour (and something weird happens for contours with level > 4):
![image](https://user-images.githubusercontent.com/18097859/81694218-5c547980-9461-11ea-9cc0-663d4c4c2f0d.png)

From my understanding, in this case I should still have labels every second contour, so for contour values of 1, 2, 3 and so on. Would you agree @UndisputedBasement33 ? Or am I misunderstanding something about this series, having never worked with it?

UndisputedBasement33 commented:
Thanks for the comment. I fixed the problem with the contours for levels > 4. I think the code is better now. I wanted too achieve a "loose" matching that's why I also added one more example with ContourLevelStep of 0.33. That means if you define LabelStep = 2 but the contour level is not exactly 2 but maybe 1.98 or 2.02 it should still show the label.

I changed the description for LabelStep, but I honestly struggle a bit to describe it precisly. Before I was relying on the Matlab describtion, which was - like you correctly mentioned - misleading. The feature should be analogue to the ContourLevelStep. With ContourLevelStep you define how many contours you want to have by defining the interval between the levels. And with LabelStep you define the interval between the levels for the labels. For example LabelStep 2 means only contours with a level which is an integral multiple of 2 will get a label. Do you understand now?

Should I again squash the commits or is it better for you to review, if I leave it like it is now?

Jonarw commented:
Thanks for the explanation, I think I understand better now how this works.
I do however still have some concerns. What if the user for some reason wants to use a non-integer `ContourLevelStep`? With `ConturLevelStep==Math.PI/2` and `LabelStep==1` I get no labels at all, except at 0:
![image](https://user-images.githubusercontent.com/18097859/81728168-52933c00-948a-11ea-9051-b26e1a41cbdb.png)

I see two possibilities how this could be solved:
1. Let the `LabelStep` not define an interval in data space, rather let it just count contours. So e.g. for `LabelStep==2`, show a label on every second contour, regardless of the `ContourLevelStep`
2. Make the `LabelStep` a double and allow fractional label steps

Personally I think I would prefer possibility 1, as it seems more intuitive to me. But as I said I do not have too much experience with this kind of series, so maybe this would be different to others.

UndisputedBasement33 commented:
I understand your concerns. I used the first option at the beginning, and it worked nicely if the contours are next to each other. But since the foreach-loop loops more or less randomly through the contours it wasn't necessarily the case that the next contour in the loop is also the contour that the user would define as next contour in the plot. Because of that out of every 2nd, it became a bit random.
But now I think I found a better way. I went back with the definition to what @Jonarw understood at first and what he suggested with his 1st option. Now `LabelStep==2` means every 2nd contour, `LabelStep==3` every 3rd contour etc..
Please let me know what you think about it

P.S: I have a failed check (AppVeyor was unable to build non-mergeab...). For me the solution builds fine, what can I do to fix that?

UndisputedBasement33 commented:
I fixed the issues commented by @Jonarw . I squashed all commits and rebased it on the latest develop branch. I hope everything is done correctly. No all the checks have passed.

Jonarw commented:
Thanks for taking care of this! 
Looks mostly good to me now, I just noticed there seems to be a slight issue with the `MultiLabel` algorithm. When I zoom in on one of the contours, it looks like this:
![image](https://user-images.githubusercontent.com/18097859/81945813-d40bdb00-95fe-11ea-861a-94a3c0619dd5.png)

UndisputedBasement33 commented:
Yes this wasn't nice, glad you found it. Took me a bit to figure out, but it should be fixed now.

Jonarw commented:
Looks good to me now!
